src/wayland/buffer/mod.rs

```diff
@@ -0,0 +1,164 @@
+//! Buffer management utilities.
+//!
+//! This module provides the [`Buffer`] type to represent a [`WlBuffer`] managed by Smithay and data
+//! associated with said buffer. This module has a dual purpose. This module provides a way for the compositor
+//! to be told when a client has destroyed a buffer. The other purpose is to provide a way for specific types
+//! of [`WlBuffer`] to abstractly associate some data with the protocol object without conflicting
+//! [`Dispatch`](crate::reexports::wayland_server::Dispatch) implementations.
+//!
+//! Unlike most delegate types found in other modules of the wayland frontend, this module has no `delegate`
+//! macro by design.
+
+use std::{any::Any, sync::Arc};
+
+use wayland_server::{
+    backend::{protocol::Message, ClientId, Handle, InvalidId, ObjectData, ObjectId},
+    protocol::wl_buffer::WlBuffer,
+    DataInit, DisplayHandle, New, Resource,
+};
+
+use crate::utils::UnmanagedResource;
+
+/// Handler trait for associating data with a [`WlBuffer`].
+///
+/// This trait primarily allows compositors to be told when a buffer is destroyed.
+///
+/// # For buffer abstractions
+///
+/// Buffer abstractions (such as [`shm`](crate::wayland::shm)) should require this trait bound in dispatch
+/// requirements to ensure access to the [`BufferManager`].
+pub trait BufferHandler {
+    /// Returns a reference to the buffer manager.
+    fn buffer_manager(&self) -> &BufferManager;
+
+    /// Called when the client has destroyed the buffer.
+    ///
+    /// At this point the buffer is no longer usable by Smithay.
+    fn buffer_destroyed(&mut self, buffer: &ManagedBuffer);
+}
+
+/// The buffer manager.
+///
+/// This type allows a [`WlBuffer`] to have data associated and managed by Smithay.
+#[derive(Debug)]
+pub struct BufferManager(());
```

Comment:
If we want to keep this type, wouldn't this be cleaner?
```suggestion
pub struct BufferManager;
```